### PR TITLE
65461: support multiple keys in the client keystore (select the matching one by keyType)

### PR DESCRIPTION
**Pull request by veselov**
## WiP STATUS
- there don't seem to be any tests around the key manager at present, please see comment below
- I'm not sure if this needs any documentation change, I don't think so, as it's just a bug fix.

## Description
Fixes support for cases when the key type request to the key manager
does not match the key type of the available keys. Without this fix,
a key of incorrect type may be returned to the SSL engine, resulting
in key misuse and connection failure.

## Motivation and Context
https://bz.apache.org/bugzilla/show_bug.cgi?id=65461

## How Has This Been Tested?
The tests pass, and I tested this for my case specifically, but I don't see
any unit test that exercise this code. Any pointers or suggestions on how
to test this during build? The simplest would probably be to have a keystore
and simulate calls to the key manager.

## Types of changes
- Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [X] My code follows the [code style][style-guide] of this project.
- [  ] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

**Comment by vlsi** [comment]
I did try to review this, and my conclusion is as follows:
1) I would like to decline the patch, as adding more and more stuff to JMeter's SSL makes the stuff harder and harder to follow
2) There are no tests to cover the change. I agree SSL-related tests was not there, however, the code is already hard to follow
3) I'm inclined that `JmeterKeyStore` should be **removed**, and we should use the default `KeyStore` implementation

**Comment by vlsi** [comment]
@veselov , WDYT?

**Comment by vlsi** [comment]
We could probably use WireMock for testing: https://github.com/Quuxplusone/BugzillaToGithub

**Comment by veselov** [comment]
1. Without this fix, we can't use client-side client authentication testing, i.e., there is no workaround for bug 65461 (that I could find).
2. I can try adding wiremock tests. The examples at the link you gave look promising and fitting.
3. `JmeterKeyStore` is not a key store. It can be removed, as a class, but all the logic of handling alias selection is still needed, and will have to be moved to `JsseSSLManager`. I'm not sure this will be any better (but see 4)
4. Ah, are you suggesting to, instead of using a singleton SSL manager, have an SSL manager (and therefore individual key stores) for each thread? Thus eliminating the need for `JmeterKeyStore` and all the complexities dealing with a shared key store? That sounds fine to me. IDK how the multiple key stores are to be specified, .ZIP file?
5. I certainly don't mind if they were gone. A workaround to providing indices into an existing key store is always to copy the key store, leaving only the items that somebody cares about. I imagine this feature was done anticipating running multi-node tests, where each node gets a "range" of keys from the same keystore. It's more feasible to have different keystores for that scenario anyway.

**Comment by vlsi** [comment]
> It can be removed, as a class, but all the logic of handling alias selection is still needed, and will have to be moved to JsseSSLManager

I guess the only logic we need is to "try using jmeterproperty-based key first", which we could place into `KeyManager`.
On the other hand, I guess, the default implementation of `KeyManager` should be capable enough to select the key that supports the needed algorithm.

**Comment by vlsi** [comment]
> Ah, are you suggesting to, instead of using a singleton SSL manager

A bit of a story is that currently JMeter relies on Java's default keystore properties, and Java provides no way to configure multiple keystores to be used at the same time.

Frankly speaking, I do not have a need for multiple different keystores to be used for different thread groups, and I do not suggest implementing it.

File: src/core/src/main/java/org/apache/jmeter/util/keystore/JmeterKeyStore.java [hunk 1]
```diff
@@ -303,36 +323,60 @@ private boolean isIndexInConfiguredRange(int index) {
      * @throws IllegalArgumentException if {@link #clientCertAliasVarName}
      *                                  is not empty and no key for this alias could be found
      */
-    public String getAlias() {
+    public String getAlias(String [] keyTypes) {
```
Comment (pmouawad) [inline src/core/src/main/java/org/apache/jmeter/util/keystore/JmeterKeyStore.java:326]: I would suggest returning Optional<String> since we're touching this class

File: src/core/src/main/java/org/apache/jmeter/util/keystore/JmeterKeyStore.java [hunk 2]
```diff
@@ -303,36 +323,60 @@ private boolean isIndexInConfiguredRange(int index) {
...
+
+    }
+
+    /**
+     * Get the next, only or null alias for the specified key type.
+     *
+     * @param keyType key type that the key under the alias must have
+     * @return the next, only, or null alias
+     */
+    public String getAlias(String keyType) {
```
Comment (pmouawad) [inline src/core/src/main/java/org/apache/jmeter/util/keystore/JmeterKeyStore.java:345]: Same here regarding use of Optional

File: src/core/src/main/java/org/apache/jmeter/util/keystore/JmeterKeyStore.java (hunk 1 above)
Comment (vlsi) [reply src/core/src/main/java/org/apache/jmeter/util/keystore/JmeterKeyStore.java:326]: this class should be removed though. I don't think it adds value :-/

Comment (pmouawad) [reply src/core/src/main/java/org/apache/jmeter/util/keystore/JmeterKeyStore.java:326]: As @veselov  explained, it implements the ability to test client certificate authentication.
This feature is useful and needed, so maybe it requires refactoring but just removing it is not acceptable.

**Comment by FSchumacher** [comment]
To get this going again.
* To merge this, tests would be needed.
* Vladimir proposes to use Java's Keystore instead. Would this be feasible without breaking backwards compatibility? Is there a good way to introduce this in small steps (with a feature flag, for example)?
* Can the API changes be made without breaking old clients? (E.g. without the removal of public methods)

**Comment by veselov** [comment]
It's most likely possible to implement this using keystores directly. This won't help with the testing, as there are still no tests around it.
It can be implemented such that the new implementation is completely separate, so no APIs are touched, and then there is a switch somewhere that selects old or new implementation.
